ping.py

```diff
@@ -0,0 +1,50 @@
+import time
+
+from sqlalchemy.engine.url import make_url
+
+from CTFd.config import Config
+
+url = make_url(Config.DATABASE_URL)
+url.database = None
+
+
+class PingError(Exception):
+    pass
+
+
+def mysql_ping(**kwargs):
+    """this imitates `mysqladmin ping`"""
+    from pymysql import Error, connect
+
+    try:
+        connect(**kwargs).ping()
+    except Error:
+        raise PingError()
+
+
+def postgres_ping(**kwargs):
+    """NotImplemented"""
+
+
+print(f"Waiting for {url} to be ready")
+
+while True:
+    try:
+        if url.drivername.startswith("mysql"):
+            mysql_ping(
+                host=url.host,
+                user=url.username,
+                password=url.password,
+                port=url.port,
+                database=None,
+            )
+        elif url.drivername.startswith("postgres"):
+            postgres_ping()
```

Comment:
This is the wrong strategy to go with. You just need to do a `raw_connection()` on the URL without the database. Separating it out by postgres and mysql is unnecessary. You only really need to special case sqlite. For sqlite you can just `exit(0)`

Reply:
`raw_connection` for postgres doesn't support `ping`ing

Reply:
I don't think you need to `ping()` specifically. `raw_connection()` in my testing worked only when the database was available for a connection. Otherwise it returned errors.

Reply:
yes, it also checks if a database is created. this isn't what `mysqladmin ping` does.

Reply:
mysqladmin ping just connects to the server. It doesn't test for database existence. 

https://serverfault.com/questions/32459/monitoring-what-does-mysqladmin-ping-report
https://dev.mysql.com/doc/refman/8.0/en/mysqladmin.html

Reply:
raw_connection doesn't need a database specified, you just need to remove it from the URL.

Reply:
> mysqladmin ping just connects to the server. It doesn't test for database existence.
> 
> [serverfault.com/questions/32459/monitoring-what-does-mysqladmin-ping-report](https://serverfault.com/questions/32459/monitoring-what-does-mysqladmin-ping-report)
> [dev.mysql.com/doc/refman/8.0/en/mysqladmin.html](https://dev.mysql.com/doc/refman/8.0/en/mysqladmin.html)

I was trying to indicate the same thing... I wasn't aware that `raw_connection` accepts None as a database name...

Reply:
do you have some instant messaging app? it's is so inefficient making a discussion here......